server/src/main/java/org/candlepin/dto/api/v1/PoolTranslator.java

```diff
@@ -122,33 +132,74 @@ public PoolDTO populate(ModelTranslator modelTranslator, Pool source, PoolDTO de
                 dest.setBranding(Collections.emptySet());
             }
 
-            Set<Product> products = source.getProvidedProducts();
-            if (products != null && !products.isEmpty()) {
-                for (Product prod : products) {
-                    if (prod != null) {
-                        dest.addProvidedProduct(
-                            new PoolDTO.ProvidedProductDTO(prod.getId(), prod.getName()));
+            Collection<Product> products =
+                source.getProduct() != null ? source.getProduct().getProvidedProducts() : null;
+            Set<PoolDTO.ProvidedProductDTO> providedProductDTOs = new HashSet<>();
+            addProvidedProducts(products, providedProductDTOs);
+            dest.setProvidedProducts(providedProductDTOs);
+
+            Collection<Product> derivedProducts =
+                source.getDerivedProduct() != null ? source.getDerivedProduct().getProvidedProducts() : null;
+            Set<PoolDTO.ProvidedProductDTO> derivedProvidedProductDTOs = new HashSet<>();
+            addProvidedProducts(derivedProducts, derivedProvidedProductDTOs);
+            dest.setDerivedProvidedProducts(derivedProvidedProductDTOs);
+
+            // accumulate provided products for stacked entitlements
+            if (PoolType.STACK_DERIVED.equals(source.getType())) {
```

Comment:
This is something we discussed for a bit before, and it didn't seem right to me. I wasn't familiar with the code in StackedSubPoolValueAccumulator and how it's used, and it's hard to go through, but I tried to look into it a bit, and from what I understand this is what happens: 

When we want to update sub-pools (stack derived pools) of a master pool, we find the entitlements from the same stack of that sub-pool, then find the pools off of those entitlements, and see if any of their provided products changed, and if they did change, we set those provided products on that sub-pool [here](https://github.com/candlepin/candlepin/blob/93ff253c81fe5618eeaa823e5e0a24531f34a469/server/src/main/java/org/candlepin/policy/js/pool/PoolRules.java#L577), and eventually persist it.

I assume the problem that arose that prompted you to add this change in the Translator is that those updated provided products are not reflected in the pool's marketing product, so we add them in the DTO here due to possible spec test failures? Would it not be preferable to change the code in PoolRules instead, to add those updated provided products on the marketing product directly and make sure our model data is consistent? Then we wouldn't need to do this on the DTO?

 @abhiskum Maybe I'm misunderstanding something since I'm not familiar with this part of the application? @Ceiu do you happen to know more about this?

Reply:
This probably isn't something we want to change internally in pool rules. As far as I understood the intent of this particular block, it's that we want to maintain the old flat representation of provided products on a given pool at the API level. But apparently stacks share provided products? That bit seems wrong to me, but will require some investigation to verify. If that's truly how stacks work, then this logic is fine (though maybe some other details need to be sorted out).

Reply:
This is something we should discuss. @abhiskum will set up a meeting.

Reply:
As per discussion with @barnabycourt, @Ceiu, @nikosmoum and @wolfdale, we came up with the conclusion that we do not need to accumulate provided products in case of stacked entitlement as marketing product and stack id has one to one relationship i.e. different marketing product will not have the same stacking id and all the required provided products will be available in marketing product itself from adapter.